File: app/src/solana/api/hunt.js
import { web3, BN } from '@project-serum/anchor'
import { TokenInstructions } from '@project-serum/serum'

import { 
  NATIVE_MINT,
} from "@/lib/spl-token";

import {
  getAtAAccountInfo
} from './sol'
import { useAnchor } from '../useAnchor'
import { Hunt, Player } from '../model/hunt'
import { developerPublicKey } from './developer'

export const all = async (filter = []) => {
  const { program } = useAnchor()
  const hunts = await program.value.account.huntAccount.all(filter)
  return hunts
      .sort((a, b) => {
        return b.account.timestamp - a.account.timestamp
      })
      .map(({ publicKey, account }) => {
        return new Hunt(publicKey, account)
      })
}

export async function createHunt(
    qsAccouts, 
    amount,
  ) {
  const { program, wallet } = useAnchor()
  const ata = await getAtAAccountInfo()
  const newHuntAccount = web3.Keypair.generate()
  console.log(qsAccouts, new BN(amount))
  await program.value.rpc.createHunt(
    qsAccouts,
    new BN(amount),
    {
      accounts: {
        huntAccount: newHuntAccount.publicKey,
        signer: wallet.value.publicKey,
        systemProgram: web3.SystemProgram.programId,

        authority: wallet.value.publicKey,
        mint: NATIVE_MINT,
        from: ata.publicKey,
        to: developerPublicKey,
        tokenProgram: TokenInstructions.TOKEN_PROGRAM_ID
      },
      signers: [ newHuntAccount ]
    }
  );
}

export async function joinHunt(huntAccount) {
  const { program, wallet } = useAnchor()
  const ata = await getAtAAccountInfo()
  const newPlayerAccount = web3.Keypair.generate()
  await program.value.rpc.joinHunt(
    {
      accounts: {
        playerAccount: newPlayerAccount.publicKey,
        huntAccount: huntAccount,
        signer: wallet.value.publicKey,
        systemProgram: web3.SystemProgram.programId,

        authority: wallet.value.publicKey,
        mint: NATIVE_MINT,
        from: ata.publicKey,
        to: developerPublicKey,
        tokenProgram: TokenInstructions.TOKEN_PROGRAM_ID
      },
      signers: [ newPlayerAccount ]
    }
  );
  return await fetchPlayer(newPlayerAccount.publicKey)
}

export async function fetchHunt(publicKey) {
  const { program } = useAnchor()
  return new Hunt(publicKey, await program.value.account.huntAccount.fetch(publicKey))
}

export async function fetchPlayer(publicKey) {
  const { program } = useAnchor()
  return new Player(publicKey, await program.value.account.huntPlayerAccount.fetch(publicKey))
}

export async function approveHunt() {
}

File: programs/solana-dapp-template/src/qa/instruction/hunt.rs
use anchor_lang::prelude::*;
use anchor_spl::token::{ Transfer };

use crate::qa::state::*;

#[derive(Accounts)]
pub struct CreateHunt<'info> {
  #[account(
    init,
    payer = signer, 
    space = HuntAccount::LENGTH
  )]
  pub hunt_account: Account<'info, HuntAccount>,
 
  #[account(mut)]
  pub signer: Signer<'info>,
  pub system_program: Program<'info, System>,

  #[account(signer)]
  pub authority: AccountInfo<'info>,
  #[account(mut)]
  pub from: AccountInfo<'info>,
  #[account(mut)]
  pub to: AccountInfo<'info>,
  pub token_program: AccountInfo<'info>,
}

impl<'a, 'b, 'c, 'info> From<&mut CreateHunt<'info>>
  for CpiContext<'a, 'b, 'c, 'info, Transfer<'info>> 
{
  fn from(accounts: &mut CreateHunt<'info>) -> CpiContext<'a, 'b, 'c, 'info, Transfer<'info>> {
    let cpi_accounts = Transfer {
      authority: accounts.authority.clone(),
      from: accounts.from.clone(),
      to: accounts.to.clone(),
    };
    let program = accounts.token_program.clone();
    CpiContext::new(program, cpi_accounts)
  }
}

#[derive(Accounts)]
pub struct JoinHunt<'info> {
  #[account(
    init,
    payer = signer, 
    space = HuntPlayerAccount::LENGTH
  )]
  pub player_account: Account<'info, HuntPlayerAccount>,
  #[account(mut)]
  pub signer: Signer<'info>,
  pub system_program: Program<'info, System>,
  #[account(mut)]
  pub hunt_account: Account<'info, HuntAccount>,
 
  #[account(signer)]
  pub authority: AccountInfo<'info>,
  #[account(mut)]
  pub from: AccountInfo<'info>,
  #[account(mut)]
  pub to: AccountInfo<'info>,
  pub token_program: AccountInfo<'info>,
}

impl<'a, 'b, 'c, 'info> From<&mut JoinHunt<'info>>
  for CpiContext<'a, 'b, 'c, 'info, Transfer<'info>> 
{
  fn from(accounts: &mut JoinHunt<'info>) -> CpiContext<'a, 'b, 'c, 'info, Transfer<'info>> {
    let cpi_accounts = Transfer {
      authority: accounts.authority.clone(),
      from: accounts.from.clone(),
      to: accounts.to.clone(),
    };
    let program = accounts.token_program.clone();
    CpiContext::new(program, cpi_accounts)
  }
}

#[derive(Accounts)]
pub struct AnswerHunt<'info> {
  #[account(mut)]
  pub builder_account: Account<'info, HuntAccount>,
  #[account(mut)]
  pub ata: AccountInfo<'info>,
}

#[derive(Accounts)]
pub struct ApproveHunt<'info> {
  #[account(mut)]
  pub hunt_account: Account<'info, HuntAccount>,
 
  #[account(signer)]
  pub authority: AccountInfo<'info>,
  #[account(mut)]
  pub from: AccountInfo<'info>,
  #[account(mut)]
  pub to: AccountInfo<'info>,
  pub token_program: AccountInfo<'info>,
}

impl<'a, 'b, 'c, 'info> From<&mut ApproveHunt<'info>>
  for CpiContext<'a, 'b, 'c, 'info, Transfer<'info>> 
{
  fn from(accounts: &mut ApproveHunt<'info>) -> CpiContext<'a, 'b, 'c, 'info, Transfer<'info>> {
    let cpi_accounts = Transfer {
      authority: accounts.authority.clone(),
      from: accounts.from.clone(),
      to: accounts.to.clone(),
    };
    let program = accounts.token_program.clone();
    CpiContext::new(program, cpi_accounts)
  }
}
